File: backend/scheduling/internal/middleware/logging.go
package middleware

import (
	"net/http"
	"strings"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"

	"github.com/maksmelnyk/scheduling/internal/logger"
)

type statusWriter struct {
	http.ResponseWriter
	statusCode int
}

func (w *statusWriter) WriteHeader(code int) {
	w.statusCode = code
	w.ResponseWriter.WriteHeader(code)
}

func getUserIdFromToken(authHeader string) string {
	if authHeader != "" {
		tokenString := strings.TrimPrefix(authHeader, "Bearer ")
		token, _, err := new(jwt.Parser).ParseUnverified(tokenString, jwt.MapClaims{})
		if err == nil {
			if claims, ok := token.Claims.(jwt.MapClaims); ok {
				userID, _ := claims["sub"].(string)
				return userID
			}
		}
	}

	return ""
}

func LoggingMiddleware(log *logger.AppLogger) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()
			sw := &statusWriter{ResponseWriter: w, statusCode: http.StatusOK}

			midLogger := log.With(
				logger.Field{Key: "request_id", Value: uuid.New().String()},
				logger.Field{Key: "http_method", Value: r.Method},
				logger.Field{Key: "http_path", Value: r.URL.Path},
				logger.Field{Key: "http_query", Value: r.URL.RawQuery},
				logger.Field{Key: "client_ip", Value: r.RemoteAddr},
				logger.Field{Key: "user_id", Value: getUserIdFromToken(r.Header.Get("Authorization"))},
			)

			r = r.WithContext(logger.WithLogger(r.Context(), midLogger))

			next.ServeHTTP(sw, r)

			duration := time.Since(start)

			midLogger = midLogger.With(
				logger.Field{Key: "status_code", Value: sw.statusCode},
				logger.Field{Key: "duration_ms", Value: duration.Milliseconds()},
			)

			if sw.statusCode >= 500 {
				midLogger.Error("Request completed with server error")
			} else if sw.statusCode >= 400 {
				midLogger.Warn("Request completed with user error")
			} else {
				midLogger.Info("Request completed successfully")
			}
		})
	}
}

File: backend/payment/app/middlewares/logging.py
import uuid
import time

from jose import jwt
from typing import Optional, Dict, Any
from loguru import logger
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.responses import Response


def get_user_id_from_token(auth_header: Optional[str]) -> Optional[str]:
    if not auth_header or not auth_header.startswith("Bearer "):
        return None

    try:
        token: str = auth_header.split(" ")[1]
        payload: Dict[str, Any] = jwt.decode(
            token=token, key="", options={"verify_signature": False}
        )
        return payload.get("sub")
    except Exception:
        return None


class RequestLoggingMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ) -> Response:
        auth_header: str | None = request.headers.get("Authorization")
        start_time: float = time.time()

        with logger.contextualize(
            request_id=str(object=uuid.uuid4()),
            http_method=request.method,
            http_path=request.url.path,
            http_query=request.url.query,
            client_ip=request.client.host if request.client else "anonymous",
            user_id=get_user_id_from_token(auth_header=auth_header),
        ):

            duration: float = time.time() - start_time

            response: Response = await call_next(request)

            log_data: Dict[str, Any] = {
                "status_code": response.status_code,
                "duration_ms": duration,
            }

            if response.status_code >= 500:
                logger.error("Request completed with server error", extra=log_data)
            elif response.status_code >= 400:
                logger.warning("Request completed with user error", extra=log_data)
            else:
                logger.info("Request completed successfully", extra=log_data)

            return response
